File: javascript/medium/meetingRoomsII.js
/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {number}
 */
var minMeetingRooms = function(intervals) {
    const startIntervals = new Array(), endIntervals = new Array();
    for(let interval of intervals) {
        startIntervals.push(interval.start);
        endIntervals.push(interval.end);
    }
    startIntervals.sort(function(x, y) { return x-y; });
    endIntervals.sort(function(x, y) { return x-y; });
    let i = 0, j = 0, count = 0, minRooms = 0;
    while(i < startIntervals.length && j < endIntervals.length) {
        if(startIntervals[i] < endIntervals[j]) {
            count++; i++;
        } else {
            count--; j++;
        }
        minRooms = Math.max(minRooms, count);
    }
    return minRooms;
};

File: ruby/medium/meeting_rooms_ii.rb

# Definition for an interval.
# class Interval
#     attr_accessor :start, :end
#     def initialize(s=0, e=0)
#         @start = s
#         @end = e
#     end
# end

# @param {Interval[]} intervals
# @return {Integer}
def min_meeting_rooms(intervals)
    return 0 if intervals.empty?
    start_intervals, end_intervals = Array.new, Array.new
    intervals.each do |interval|
        start_intervals.push(interval.start)
        end_intervals.push(interval.end)
    end
    start_intervals.sort!
    end_intervals.sort!
    i, j, num_rooms, min_rooms = 0, 0, 0, 0
    while i < start_intervals.length && j < end_intervals.length
        if start_intervals[i] < end_intervals[j]
            num_rooms += 1
            i += 1
        else
            num_rooms -= 1
            j += 1
        end
        min_rooms = [min_rooms, num_rooms].max
    end
    return min_rooms
end
